### PR TITLE
Carabus 2.0

### PR DESCRIPTION
**Pull request by DaniilSokolyuk**
- [x] Package RDS.CaraBus.InMemory removed and moved to RDS.CaraBus package
- [x] Simplified configuration by using InMemoryCaraBusOptions and RabbitMQCaraBusOptions
- [x] RDS.CaraBus.InMemory.Carabus renamed to RDS.CaraBus.InMemory.InMemoryCaraBus
- [x] RDS.CaraBus.RabbitMQ.Carabus renamed to RDS.CaraBus.RabbitMQ.RabbitMQCaraBus
- [x] Added a task queue implementation for prevent thread pool exhaustion & extremely bad performance
- [x] Removed MaxConcurrentHandlers from SubscribeOptions
- [x] Added MaxDegreeOfParallelism option, (default - 4)
- [x] Added CancellationToken everywhere
- [x] Truly async
- [x] Added AutoStart option, (default - false), if true - not needed call StartAsync()
- [x] Graceful shutdown with StopAsync (or dispose synchronously)

### CODE REVIEW

File: src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                {
+                    if (_publishChannel == null)
+                    {
+                        _publishChannel = _connection.CreateModel();
+                    }
+                }
+            }
+        }
+
+        public override async Task SubscribeAsyncImpl(Type messageType, Func<object, CancellationToken, Task> handler, SubscribeOptions subscribeOptions, CancellationToken cancellationToken)
```
Comment (sergey-borisenko) [inline src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs]: protected

Comment (DaniilSokolyuk) [reply src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs]: fixed

File: src/RDS.CaraBus/CaraBusBase.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            _subscribeActions.Add(SubFunc);
+
+            return Task.CompletedTask;
+        }
+
+        protected abstract Task SubscribeAsyncImpl(Type messageType, Func<object, CancellationToken, Task> handler, SubscribeOptions subscribeOptions, CancellationToken cancellationToken);
+
+        public async Task StartAsync()
```
Comment (sergey-borisenko) [inline src/RDS.CaraBus/CaraBusBase.cs:68]: Create -> Stop -> Start

Comment (DaniilSokolyuk) [reply src/RDS.CaraBus/CaraBusBase.cs]: fixes

Comment (DaniilSokolyuk) [reply src/RDS.CaraBus/CaraBusBase.cs:68]: fixed

File: src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        public RabbitMQCaraBus(RabbitMQCaraBusOptions rabbitMQCaraBusOptions, ILoggerFactory loggerFactory = null)
+            : base(
+                rabbitMQCaraBusOptions,
+                new TaskQueue(maxItems: 100, maxDegreeOfParallelism: rabbitMQCaraBusOptions.MaxDegreeOfParallelism, loggerFactory: loggerFactory), 
+                loggerFactory)
+        {
+            _connectionFactory = new ConnectionFactory
+            {
+                Uri = new Uri(rabbitMQCaraBusOptions.ConnectionString),
+                DispatchConsumersAsync = true
```
Comment (Toxu4) [inline src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs:44]: :+1:

File: src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                }
+            }
+        }
+
+        protected override async Task SubscribeAsyncImpl(Type messageType, Func<object, CancellationToken, Task> handler, SubscribeOptions subscribeOptions, CancellationToken cancellationToken)
+        {
+            await EnsureConnectionAndPublishChannelCreated().ConfigureAwait(false);
+
+            var exchangeName = GetExchangeName(subscribeOptions.Scope, messageType);
+            var queueName = GetQueueName(messageType, subscribeOptions);
```
Comment (Toxu4) [inline src/RDS.CaraBus.RabbitMQ/RabbitMQCaraBus.cs]: change signature - swap arguments:

`GetQueueName(subscribeOptions, messageType);`

File: src/RDS.CaraBus/ICarabusStopAsync.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading.Tasks;
+
+namespace RDS.CaraBus
+{
+    interface ICarabusStopAsync
```
Comment (Toxu4) [inline src/RDS.CaraBus/ICarabusStopAsync.cs]: `ICaraBusStopAsync`

File: src/RDS.CaraBus/Utility/TypeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+        private static IReadOnlyCollection<Type> GetInheritanceChainAndInterfacesImpl(Type type)
+        {
+            var chain = new List<Type>();
+
+            chain.AddRange(type.GetInterfaces());
+
+            while (true)
+            {
+                if (type == null)
```
Comment (Toxu4) [inline src/RDS.CaraBus/Utility/TypeExtensions.cs]: wtf?

`while (type != null)`

**Review by Toxu4 — CHANGES_REQUESTED** [review]

**Pull request merged**